### DISCUSSION TITLE
How to use container packaging support with Typescript and the serverless-esbuild plugin

### DISCUSSION THREAD

**Discussion by reikje** [serverless/serverless#11119] [Q&A]
I have a project which is based on the latest `aws-nodejs-typescript` template. The directory layout is basically the same as mentioned [here](https://blog.logrocket.com/building-serverless-app-typescript/). Unfortunately, it turned out that I need to package it as a docker container because I need to run on Ubuntu and install an extra binary.

While I have found a number of examples for JavaScript, I'd like to stay on TypeScript. Is there an example of a `Dockerfile` that uses the output from the `serverless-esbuild` plugin when the docker image is created? Does the `ecr` step run after the `serverless-esbuild` plugin, so I can access the transpiled JavaScript files?

**Comment by pgrzesik** [comment]
Hey @reikje - I would personally just run `esbuild` normally - not via a plugin, as you're controlling the whole build process via your `Dockerfile`. Plugins like `serverless-esbuild` just ignore image-based functions (https://github.com/floydspace/serverless-esbuild/blob/1cb893fbdd02f991f749959571c3f6045195c77b/src/index.ts#L154-L158) because they don't have handler and basically anything can happen inside the docker image you're preparing.

**Reply by reikje** [reply]
Hi @pgrzesik yes that's one option, essentially doing [this](https://adieuadieu.medium.com/using-esbuild-in-your-serverless-framework-project-13723db5e32a). I want to give the `serverless-esbuild` plugin one more shot. The package is simply a zip file that end's up in `.serverless`, so in theory I should be able to unzip and use that within my `Dockerfile` if the timing allows it.

**Reply by pgrzesik** [reply]
You might want to try that, but I'm not sure if `serverless-esbuild` plugin will even run as it filters out the functions that do not have handler and node runtime. If that works, please report back as someone might find that information useful in the future 💯

**Reply by reikje** [reply]
@pgrzesik you are correct, it simply excludes functions that do not have a handler.

> this plugin will automatically ignore any function which does not contain a handler

`Packaging: No files found. Skipping esbuild.`

I created a [ticket](https://github.com/floydspace/serverless-esbuild/issues/316) over at the `serverless-esbuild` plugin for this, let's see what happens.

As an alternative, maybe there is another way to transpile Typescript to JavaScript during `sls deploy`, even if packaging isn't happening? Do you know any other plugins or integration hooks that I could utilize instead?

**Reply by pgrzesik** [reply]
Hey @reikje 👋  I'm not aware of any plugins that do that - you would have to do it on your own with custom plugin or writing custom logic with something like https://github.com/weixu365/serverless-scriptable-plugin. Not sure if that is better than moving the packaging process to `docker build` though.

**Reply by reikje** [reply]
@pgrzesik ah that's a great idea. I can basically instead copy the `.ts` files into the container and transpile as part of `docker build` - haven't even thought of that. Can I ask one last thing in regards to the following from the [docs](https://www.serverless.com/framework/docs/providers/aws/guide/functions):

> Serverless will create an ECR repository for your image, but it currently does not manage updates to it. An ECR repository is created only for new services or the first time that a function configured with an image is deployed. In service configuration, you can configure the ECR repository to scan for CVEs via the provider.ecr.scanOnPush property, which is false by default. 

Does this mean, if I change my existing function code, run `sls deploy` to deploy updates - that I won't see any update in my Lambda functions?

**Reply by pgrzesik** [reply]
You will see them - the part from docs is about provisioning the ECR repository to which you the images will be uploaded - it will be only created once and any configuration changes (right now we only have `scanOnPush` on that level) afterwards won't be applied. Images will be uploaded during each new deploy/package

**Comment by reikje** [comment]
I solved this by building manually and not using the `serverless-esbuild` plugin. Pasting my setup here as a reference. My function is in `src/functions/preview/handler.ts`. 

**Dockerfile**
```
FROM ubuntu:20.04

ENV DEBIAN_FRONTEND=noninteractive
ENV DISPLAY :99

ENV STAGE=dev
ENV REGION=us-east-2
ENV BUCKET_NAME=bucket

RUN apt-get update
RUN apt-get upgrade --yes
RUN apt-get install -y \
    autoconf \
    cmake \
    curl \
    g++ \
    openscad \
    libcurl4-openssl-dev \
    libtool \
    make \
    xvfb \
    && rm -rf /var/lib/apt/lists/*
RUN curl -sL https://deb.nodesource.com/setup_16.x | bash
RUN apt-get install nodejs --yes

COPY package.json tsconfig.json tsconfig.paths.json esbuild.config.js ./

RUN npm install -g aws-lambda-ric
RUN npm install

COPY src/ ./src/
RUN node ./esbuild.config.js
RUN cp ./dist/handler.js ./handler.js

RUN rm -rf package.json tsconfig.json tsconfig.paths.json esbuild.config.js

# Downloads the Lambda Runtime Interface Emulator (RIE)
ADD https://github.com/aws/aws-lambda-runtime-interface-emulator/releases/latest/download/aws-lambda-rie /usr/bin/aws-lambda-rie

RUN chmod +x /usr/bin/aws-lambda-rie
COPY entry.sh ./
RUN chmod +x ./entry.sh

ENTRYPOINT ["./entry.sh"]
CMD ["handler.main"]
```

**esbuild.config.js**
```
const esbuild = require('esbuild');

esbuild
  .build({
    entryPoints: ['./src/functions/preview/handler.ts'],
    outfile: 'dist/handler.js',
    bundle: true,
    minify: false,
    sourcemap: false,
    external: ['aws-sdk'],
    target: 'node16',
    define: { 'require.resolve': undefined },
    platform: 'node',
  })
  .catch(() => process.exit(1));
```


**entry.sh**
```
#!/bin/sh

# Check if the AWS_LAMBDA_RUNTIME_API is not set. This environment
# variable is set when the image is running on Lambda.
if [ -z "${AWS_LAMBDA_RUNTIME_API}" ]; then
  # We know the image is being run off of Lambda, so we need to use the RIE
  # to start the function.
  exec /usr/bin/aws-lambda-rie /usr/bin/aws-lambda-ric $1
else
  # We know the image is being run on Lambda so we don't need to use the RIE.
  exec /usr/bin/aws-lambda-ric $1
fi
```

**Reply by pgrzesik** [reply]
Thanks for sharing @reikje 🙇

**Comment by taschmidt** [comment]
Building on the workaround suggested by @reikje, this seems to be working well for me. I used the versatile [serverless-hooks plugin](https://github.com/jlopez/serverless-hooks) and added a `hook:before:package:initialize` script which runs my build. That way I don't need to manually add a call to my build script everywhere it may be needed.
